### PR TITLE
cuda improvement

### PR DESCRIPTION
**Pull request by zzzming**
Add a multiprocessor to run passages encode independently.

### CODE REVIEW

File: ragstack/colbert/colbert_embedding.py [hunk 1]
```diff
@@ -161,30 +174,14 @@ def encode_query(
...
+        launcher = Launcher(encode_passages, nranks=self.__nranks)
 
-            start_idx = end_idx
+        manager = mp.Manager()
+        shared_lists = [manager.list() for _ in range(self.__nranks)]
+        shared_queues = [manager.Queue(maxsize=1) for _ in range(self.__nranks)]
 
-        return embedded_chunks
+        # Encodes collection into index using the CollectionIndexer class
+        logging.info(f"nranks: {self.__nranks}")
```
Comment (nicoloboschi) [inline ragstack/colbert/colbert_embedding.py:184]: no need

File: ragstack/colbert/distributed.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            self._setup(nranks)
+            self._is_initialized = True
+
+    def _setup(self, nranks: int):
+        master_addr = "127.0.0.1"
+        master_port = find_free_port()
+        if nranks < 1:
+            nranks = torch.cuda.device_count()
+        world_size = min(torch.cuda.device_count(), nranks)
+        logging.info(f"setting up resource group {world_size=}")
```
Comment (nicoloboschi) [inline ragstack/colbert/distributed.py:90]:
```suggestion
        logging.info(f"setting up resource group {world_size}")
```

**Review by nicoloboschi — APPROVED** [review]
This is very cool! great work @zzzming

**Review by epinzur — APPROVED** [review]
very cool

**Pull request merged**